HW2/P2/parallel_vector.pyx

```diff
@@ -79,12 +79,18 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+
+   #with nogil:
+   for r in range(repeat):
+       for idx in prange(src.shape[0], nogil=True, num_threads=4, schedule=dynamic):
+       #for idx in range(src.shape[0]):
+           if counts[src[idx]] > 0:
```

Comment:
should acquire lock before accessing counts! because counts could be changing in the meantime!